src/concord-once.c

```diff
@@ -36,43 +43,94 @@ static void
 _ccord_sigint_handler(int signum)
 {
     (void)signum;
-    fputs("\nSIGINT: Disconnecting running concord client(s) ...\n", stderr);
-    pthread_mutex_lock(&shutdown_lock);
-    ccord_should_shutdown = 1;
-    pthread_mutex_unlock(&shutdown_lock);
+    const char *err_str =
+        "\nSIGINT: Disconnecting running concord client(s) ...\n";
+    write(STDERR_FILENO, err_str, strlen(err_str));
+    ccord_shutdown_async();
 }
 #endif /* CCORD_SIGINTCATCH */
 
 CCORDcode
 ccord_global_init()
 {
-    if (once) {
-        return CCORD_GLOBAL_INIT;
-    }
-    else {
+    int ret_code = CCORD_OK;
+    pthread_mutex_lock(&lock);
+    if (0 == init_counter++) {
 #ifdef CCORD_SIGINTCATCH
         signal(SIGINT, &_ccord_sigint_handler);
 #endif
         if (0 != curl_global_init(CURL_GLOBAL_DEFAULT)) {
             fputs("Couldn't start libcurl's globals\n", stderr);
-            return CCORD_GLOBAL_INIT;
+            goto fail_curl_init;
         }
-        if (discord_worker_global_init()) {
+        if (0 != discord_worker_global_init()) {
             fputs("Attempt duplicate global initialization\n", stderr);
-            return CCORD_GLOBAL_INIT;
+            goto fail_discord_worker_init;
+        }
+        if (0 != pipe(shutdown_fds)) {
+            fputs("Failed to create shutdown pipe\n", stderr);
+            goto fail_pipe_init;
+        }
+        for (int i = 0; i < 2; i++) {
+            const int on = 1;
+            if (0 != ioctl(shutdown_fds[i], FIOCLEX, NULL)) {
+                fputs("Failed to make shutdown pipe close on execute\n",
+                      stderr);
+                goto fail_pipe_init;
+            }
+            if (0 != ioctl(shutdown_fds[i], FIONBIO, &on)) {
+                fputs("Failed to make shutdown pipe nonblocking\n", stderr);
+                goto fail_pipe_init;
+            }
+        }
+
+        if (0) {
```

Comment:
I see why you did this, but I find it a bit weird tbh. We can write a couple more lines here:
```c
    pthread_mutex_unlock(&lock);
    return ret_code;
    
fail_pipe_init:
    for (int i = 0; i < 2; i++) {
        if (-1 != shutdown_fds[i]) {
            close(shutdown_fds[i]);
            shutdown_fds[i] = -1;
        }
     }
fail_discord_worker_init:
    discord_worker_global_cleanup();
fail_curl_init:
    curl_global_cleanup();

    ret_code = CCORD_GLOBAL_INIT;
    init_counter = 0;
            
    pthread_mutex_unlock(&lock);
    return ret_code;

Reply:
So this function only returns either `CCORD_OK` or `CCORD_GLOBAL_INIT`, we can get rid of `ret_code` and just return the immediate value